rust/lance-io/src/object_store/gcs_wrapper.rs

```diff
@@ -0,0 +1,458 @@
+                    }
+                    break;
+                }
+                UploadState::PuttingSingle(ref mut fut) | UploadState::Completing(ref mut fut) => {
+                    match fut.poll_unpin(cx) {
+                        Poll::Ready(Ok(())) => mut_self.state = UploadState::Done,
+                        Poll::Ready(Err(e)) => {
+                            return Err(std::io::Error::new(std::io::ErrorKind::Other, e))
+                        }
+                        Poll::Pending => break,
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl AsyncWrite for Upload {
+    fn poll_write(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+        buf: &[u8],
+    ) -> std::task::Poll<Result<usize, std::io::Error>> {
+        self.as_mut().poll_tasks(cx)?;
+
+        // Fill buffer up to remaining capacity.
+        let remaining_capacity = self.buffer.capacity() - self.buffer.len();
+        let bytes_to_write = std::cmp::min(remaining_capacity, buf.len());
+        self.buffer.extend_from_slice(&buf[..bytes_to_write]);
+
+        // Rust needs a little help to borrow self mutably and immutably at the same time
+        // through a Pin.
+        let mut_self = &mut *self;
+
+        // Instantiate next request, if available.
+        if mut_self.buffer.capacity() == mut_self.buffer.len() {
+            match &mut mut_self.state {
+                UploadState::Pending => {
+                    let store = self.store.clone();
+                    let path = self.path.clone();
+                    let fut = Box::pin(async move { store.create_multipart(path.as_ref()).await });
+                    self.state = UploadState::CreatingUpload(fut);
+                }
+                UploadState::InProgress {
+                    multipart_id,
+                    part_idx,
+                    futures,
+                    ..
+                } => {
+                    // TODO: Make max concurrency configurable.
+                    if futures.len() < 10 {
+                        futures.push(Self::put_next_part(
+                            &mut_self.path,
+                            &mut_self.store,
+                            &mut mut_self.buffer,
+                            part_idx,
+                            multipart_id,
+                        ));
+                    }
+                }
+                _ => {}
+            }
+        }
+
+        self.poll_tasks(cx)?;
+
+        match bytes_to_write {
+            0 => Poll::Pending,
+            _ => Poll::Ready(Ok(bytes_to_write)),
+        }
+    }
+
+    fn poll_flush(
+        mut self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), std::io::Error>> {
+        self.as_mut().poll_tasks(cx)?;
+
+        match &self.state {
+            UploadState::Pending | UploadState::Done => Poll::Ready(Ok(())),
```

Comment:
It was confusing for me here that `UploadState::Pending` did not go to `Poll::Pending`.  Maybe change `UploadState::Pending` to `UploadState::Started` or something like that?

Reply:
Changing the name to `Started`.